File: 389.Find-the-Difference.py
# https://leetcode.com/problems/find-the-difference/description/
#
# algorithms
# Easy (52.06%)
# Total Accepted:    124.1k
# Total Submissions: 238.4k

class Solution(object):
    def findTheDifference(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        length = len(t)
        bit = 0
        
        for i in xrange(length):
            bit ^= ord(t[i])
            if i == length - 1:
                break
            bit ^= ord(s[i])
        
        return chr(bit)
        
        

File: 389.Find-the-Difference.java
// https://leetcode.com/problems/find-the-difference/
//
// algorithms
// Easy (53.93%)
// Total Accepted: 165,145
// Total Submissions: 306,205


class Solution {
    public char findTheDifference(String s, String t) {
        int[] chNum = new int[26];

        for (char ch : s.toCharArray()) {
            chNum[ch - 'a']++;
        }

        for (char ch : t.toCharArray()) {
            int idx = ch - 'a';
            chNum[idx]--;
            if (chNum[idx] < 0) {
                return ch;
            }
        }

        return ' ';
    }
}